### PR TITLE
Skip emit of this parameters

### LINKED ISSUE #8179: [Transforms] Do not emit 'this' function parameters
For the typescript:
```ts
//// [looseThisTypeInFunctions.ts]

interface I {
    n: number;
    explicitThis(this: this, m: number): number;
}
class C implements I {
    n: number;
    explicitThis(this: this, m: number): number {
        return this.n + m;
    }
    implicitThis(m: number): number {
        return this.n + m;
    }
    explicitVoid(this: void, m: number): number {
        return m + 1;
    }
}
```

Expected no `this` parameters: 

```ts
//// [looseThisTypeInFunctions.js]

var C = (function () {
    function C() {
    }
    C.prototype.explicitThis = function (m) {
        return this.n + m;
    };
    C.prototype.implicitThis = function (m) {
        return this.n + m;
    };
    C.prototype.explicitVoid = function (m) {
        return m + 1;
    };
    return C;
}());
```

Actual with emitted `this` parameters:

```ts
//// [looseThisTypeInFunctions.js]
var C = (function () {
    function C() {
    }
    C.prototype.explicitThis = function (this, m) {
        return this.n + m;
    };
    C.prototype.implicitThis = function (m) {
        return this.n + m;
    };
    C.prototype.explicitVoid = function (this, m) {
        return m + 1;
    };
    return C;
}());
```

Fails in:
1. looseThisTypeInFunctions
2. thisTypeInFunctions
3. thisTypeInFunctionsNegative
4. contextualTyping24

### PR DESCRIPTION
**Pull request by sandersn**
Fixes #8179

### CODE REVIEW

**Comment by sandersn** [comment]
Never mind, this doesn't actually work.

**Comment by sandersn** [comment]
The baselines improve slightly -- the new emitter is better with comments, and better after really bad parse failures.

File: src/compiler/transformers/ts.ts [hunk 1]
```diff
@@ -2114,10 +2115,14 @@ namespace ts {
          * This function will be called when one of the following conditions are met:
          * - The node has an accessibility modifier.
          * - The node has a questionToken.
+         * - The node's text is "this".
          *
          * @param node The parameter declaration node.
          */
         function visitParameter(node: ParameterDeclaration) {
+            if ((node.name as Identifier).text === "this") {
```
Comment (yuit) [inline src/compiler/transformers/ts.ts]: Minor: You can check by `if ((node.name as Identifier).originalKeywordKind=== SyntaxKind.ThisKeyword)`

File: tests/baselines/reference/thisTypeInFunctions.js [hunk 2]
```diff
@@ -243,7 +243,8 @@ function implicitThis(n) {
 var impl = {
     a: 12,
     explicitVoid2: function () { return _this.a; },
-    explicitVoid1: function () { return 12; },
+    explicitVoid1: // ok, this: any because it refers to some outer object (window?)
```
Comment (yuit) [inline tests/baselines/reference/thisTypeInFunctions.js]: this seems like regression ? Should the comment be there ? In the (original)[https://github.com/Microsoft/TypeScript/pull/8197/files#diff-244ce633ddb2ae835e68b8278cbf7e32R41] it should be after `explicitVoid2`

Comment (mhegazy) [inline tests/baselines/reference/thisTypeInFunctions.js]: i have seen this in other tests. can you file a different issue for it.

File: src/compiler/binder.ts [hunk 3]
```diff
@@ -2267,6 +2267,11 @@ namespace ts {
             transformFlags |= TransformFlags.AssertTypeScript;
         }
 
+        // If the parameter's name is 'this, then it is TypeScript syntax.
```
Comment (DanielRosenwasser) [inline src/compiler/binder.ts]: 'this'

File: tests/baselines/reference/thisTypeInFunctions.js (hunk 2 above)
Comment (sandersn) [inline tests/baselines/reference/thisTypeInFunctions.js]: Sure, I'll revert this. I filed #8219 to track it.

**Comment by mhegazy** [comment]
:+1:

**Comment by sandersn** [comment]
Actually #8219 is a dupe of #8209. Thanks @yuit

**Pull request merged**